Does not build for `no_std` targets

This crate does not currently build for `no_std` targets, such as microcontrollers.

```
➜  curve25519-rs git:(master) cargo build -j1 --target thumbv7em-none-eabihf
   Compiling semver-parser v0.7.0
   Compiling cc v1.0.26
   Compiling rand_core v0.3.0
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv7em-none-eabihf` target may not be installed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: Could not compile `rand_core`.

To learn more, run the command again with --verbose.
```

It appears something in rand_core is pulling in the stdlib.

Even when disabling the default features of `rand`, then the crate fails to build:

```diff
diff --git a/Cargo.toml b/Cargo.toml
index 90a3f19..41f40ab 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,8 +13,9 @@ keywords = ["crypto", "curve25519", "rust-crypto"]
 categories = ["no-std", "algorithms"]
 repository = "https://github.com/shekohex/curve25519-rs"

-[dependencies]
-rand = "0.6.1"
+[dependencies.rand]
+version = "0.6.1"
+default-features = false

 [build-dependencies]
 cc = "1.0.26"
```

```
➜  curve25519-rs git:(master) ✗ cargo build -j1 --target thumbv7em-none-eabihf
   Compiling curve25519 v0.1.0 (/tmp/curve25519-rs)
error[E0432]: unresolved import `rand::rngs::OsRng`
  --> src/lib.rs:19:12
   |
19 | use rand::{rngs::OsRng, Error as RndError, Rng};
   |            ^^^^^^^^^^^ no `OsRng` in `rngs`

error[E0599]: no method named `to_vec` found for type `[i32; 10]` in the current scope
  --> src/lib.rs:33:20
   |
33 |         self_elems.to_vec() == other_elems.to_vec()
   |                    ^^^^^^

error[E0599]: no method named `to_vec` found for type `[i32; 10]` in the current scope
  --> src/lib.rs:33:44
   |
33 |         self_elems.to_vec() == other_elems.to_vec()
   |                                            ^^^^^^

error: aborting due to 3 previous errors

Some errors occurred: E0432, E0599.
For more information about an error, try `rustc --explain E0432`.
error: Could not compile `curve25519`.

To learn more, run the command again with --verbose.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,6 +4,10 @@ rust:
   - stable
   - beta
   - nightly
+addons:
+  apt:
+    packages:
+      - gcc-arm-none-eabi
 script: |
   cargo build --verbose &&
   cargo test  --verbose &&
@@ -12,7 +16,7 @@ script: |
   cargo test  --verbose --no-default-features &&
   cargo doc   --verbose --no-default-features &&
   rustup target add thumbv7em-none-eabihf &&
-  RUSTFLAGS="-C linker=rust-lld" cargo build --verbose --target thumbv7em-none-eabihf
+  cargo build --verbose --target thumbv7em-none-eabihf
 
 matrix:
   include:

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,9 +8,11 @@ script: |
   cargo build --verbose &&
   cargo test  --verbose &&
   cargo doc   --verbose &&
-  cargo build --no-default-features &&
+  cargo build --verbose --no-default-features &&
+  cargo test  --verbose --no-default-features &&
+  cargo doc   --verbose --no-default-features &&
   rustup target add thumbv7em-none-eabihf &&
-  cargo build --target thumbv7em-none-eabihf
+  cargo build --verbose --target thumbv7em-none-eabihf
 
 matrix:
   include:

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,6 +8,7 @@ addons:
   apt:
     packages:
       - gcc-arm-none-eabi
+      - libnewlib-arm-none-eabi
 script: |
   cargo build --verbose &&
   cargo test  --verbose &&

--- a/.travis.yml
+++ b/.travis.yml
@@ -12,7 +12,7 @@ script: |
   cargo test  --verbose --no-default-features &&
   cargo doc   --verbose --no-default-features &&
   rustup target add thumbv7em-none-eabihf &&
-  cargo build --verbose --target thumbv7em-none-eabihf
+  RUSTFLAGS="-C linker=rust-lld" cargo build --verbose --target thumbv7em-none-eabihf
 
 matrix:
   include:

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,7 +7,10 @@ rust:
 script: |
   cargo build --verbose &&
   cargo test  --verbose &&
-  cargo doc   --verbose
+  cargo doc   --verbose &&
+  cargo build --no-default-features &&
+  rustup target add thumbv7em-none-eabihf &&
+  cargo build --target thumbv7em-none-eabihf
 
 matrix:
   include:

--- a/.travis.yml
+++ b/.travis.yml
@@ -16,7 +16,7 @@ script: |
   cargo test  --verbose --no-default-features &&
   cargo doc   --verbose --no-default-features &&
   rustup target add thumbv7em-none-eabihf &&
-  cargo build --verbose --target thumbv7em-none-eabihf
+  cargo build --verbose --no-default-features --target thumbv7em-none-eabihf
 
 matrix:
   include:

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,14 +13,19 @@ keywords = ["crypto", "curve25519", "rust-crypto"]
 categories = ["no-std", "algorithms"]
 repository = "https://github.com/shekohex/curve25519-rs"
 
-[dependencies]
-rand = "0.6.1"
+[dependencies.rand]
+version = "0.6.1"
+default-features = false
 
 [build-dependencies]
 cc = "1.0.26"
 
 [dev-dependencies]
 criterion = "0.2"
+
+[features]
+std = ["rand/std"]
+default = ["std"]
 
 [[bench]]
 name = "curve25519"

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -16,7 +16,12 @@ use core::{
     cmp::{min, Eq, PartialEq},
     ops::{Add, Mul, Sub},
 };
-use rand::{rngs::OsRng, Error as RndError, Rng};
+
+#[allow(unused_imports)]
+use rand::{Error as RndError, Rng};
+
+#[cfg(feature = "std")]
+use rand::rngs::OsRng;
 
 /// Here the field is \Z/(2^255-19).
 ///
@@ -30,7 +35,7 @@ impl PartialEq for FieldElement {
     fn eq(&self, other: &FieldElement) -> bool {
         let &FieldElement(self_elems) = self;
         let &FieldElement(other_elems) = other;
-        self_elems.to_vec() == other_elems.to_vec()
+        self_elems == other_elems
     }
 }
 
@@ -2646,17 +2651,23 @@ pub fn curve25519(secret: [u8; 32], public: [u8; 32]) -> [u8; 32] {
 /// # }
 /// ```
 pub fn curve25519_sk(rand: Option<[u8; 32]>) -> Result<[u8; 32], RndError> {
-    let mut buf: [u8; 32] = [0; 32];
-
     // Fill a 32-byte buffer with random values if necessary.
     // Otherwise, use the given 32-byte value.
     let mut rand: [u8; 32] = match rand {
         Some(r) => r,
+
+        #[cfg(feature = "std")]
         None => {
             let mut rng = OsRng::new()?;
+            let mut buf: [u8; 32] = [0; 32];
             rng.fill(&mut buf);
             buf
         },
+
+        #[cfg(not(feature = "std"))]
+        None => {
+            unimplemented!()
+        }
     };
 
     // curve25519 secret key bit manip.

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -2639,6 +2639,9 @@ pub fn curve25519(secret: [u8; 32], public: [u8; 32]) -> [u8; 32] {
 /// ```rust
 /// # use self::curve25519::curve25519_sk;
 /// # use rand::Error as RndError;
+/// # #[cfg(not(feature = "std"))]
+/// # fn main() { }
+/// # #[cfg(feature = "std")]
 /// # fn main() -> Result<(), RndError> {
 /// // Let curve25519_sk generate the random 32-byte value.
 /// let sk1 = curve25519_sk(None)?;

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -18,7 +18,7 @@ use core::{
 };
 
 #[allow(unused_imports)]
-use rand::{Error as RndError, Rng};
+use rand::{Error as RndError, ErrorKind::Unavailable, Rng};
 
 #[cfg(feature = "std")]
 use rand::rngs::OsRng;
@@ -2669,8 +2669,11 @@ pub fn curve25519_sk(rand: Option<[u8; 32]>) -> Result<[u8; 32], RndError> {
 
         #[cfg(not(feature = "std"))]
         None => {
-            unimplemented!()
-        }
+            return Err(RndError::new(
+                Unavailable,
+                "Cannot generate random without Standard Library",
+            ));
+        },
     };
 
     // curve25519 secret key bit manip.

